### ISSUE TITLE
Add a config parameter to display Vulkan present mode

### ISSUE THREAD

**Issue by shmerl** [flightlessmango/MangoHud#1190]
It could be useful to see what's the current [Vulkan present mode](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPresentModeKHR.html) is, since it can be affected by a bunch of factors (like in-game vsync option) which are relatively obscure if you want to figure out what present mode it ends up being.

**Comment by flightlessmango** [comment]
I'm not sure what we would title this in the hud

**Comment by shmerl** [comment]
`present_mode` may be? Or `vk_present_mode` if you want to make it more clear it's for Vulkan specifically.

**Comment by flightlessmango** [comment]
I don't think that can fit in the hud

**Comment by shmerl** [comment]
Then simply something short like `K_PRESENT_MODE_MAILBOX_KHR` or `mailbox` will do. Those who want to see it will figure it out.

**Comment by flightlessmango** [comment]
I'm not sure we are able to do this, I don't know that we can know for certain that a specific present mode is being used

**Comment by shmerl** [comment]
Hm, there is no way to query it from the Vulkan driver?

**Comment by flightlessmango** [comment]
No we can't query it directly.
But it looks like we can ask what present modes are available for the device and trust that if it's available and we've set it that it should be used

**Comment by shmerl** [comment]
I mean let's say you change the present mode in the settings (like vsync toggle and such), if you can't query it, it won't be able to pick up the change? May be it's worth proposing some API call for Khronos for querying it? It sounds like an overkill, but on the other hand it's surprising you don't have visibility into current state from the layer.

**Comment by flightlessmango** [comment]
I don't think present mode can be changed without recreating the swapchain, so I think we're fine

**Comment by shmerl** [comment]
But what do you mean "we've set it"? I.e. there might a bunch of modes available, but it was set by the game itself.

**Comment by flightlessmango** [comment]
We're able to check what the app requested for present mode and apply the same logic to that as when we overwrite the present mode.
I've written an implementation here that hopefully works as expected 2bc323b2ed045e38089eca9a9d34d3853e608d3e and 0ab4c25343ba3f61cd738f6d85d577f6f12ce855

The opengl part might not be as well as the vulkan one

**Comment by shmerl** [comment]
Thanks, you might want to try to catch the edge condition of Vulkan adding a new mode before you are ready for it (you'll hit out of bound index without checking). May be you can display something like "unknown" in that case.

**Comment by shmerl** [comment]
Tried running `vkcube` with mangohud forcing a present mode with radv like this:

`MESA_VK_WSI_PRESENT_MODE=mailbox vkcube`

Some weird results, and mangohud reports it as FIFO :)

**Comment by flightlessmango** [comment]
If the present mode is not supported then it falls back to FIFO

**Comment by shmerl** [comment]
Why wouldn't mailbox be supported though? It's a normal present mode for a Wayland compositor to use.

**Comment by flightlessmango** [comment]
Not all GPUs and drivers support every present mode so I would start by checking if yours does

**Comment by shmerl** [comment]
I'm pretty sure radv supports mailbox fine, running vkcube without mangohud like this makes it spin very fast:

```
MESA_VK_WSI_PRESENT_MODE=mailbox VK_INSTANCE_LAYERS=VK_LAYER_MESA_overlay vkcube
```

![image](https://github.com/flightlessmango/MangoHud/assets/310927/097430ae-0f16-4819-8d83-53eaf5abea23)

And without overriding it, it syncs to max display refresh rate for me (180 Hz):

![image](https://github.com/flightlessmango/MangoHud/assets/310927/dc111bc3-135a-4afb-9dc5-944930f729a6)

**Comment by shmerl** [comment]
With mangohud it also spins very fast when mailbox is forced, so I'm pretty sure it is using mailbox, but I think it misreports it as fifo.

Also, side note - I guess mangohud is more bound than Mesa overlay, it gives less framerate when mailbox is forced, but still way way above 180 fps.

**Comment by shmerl** [comment]
All the above are with KDE Plasma 5.27.10, Wayland session btw.

**Comment by shmerl** [comment]
Oh, also found that vkcube has a Wayland version (neat) and you can pass present mode to it as a parameter:

```
vkcube-wayland --present_mode VK_PRESENT_MODE_MAILBOX_KHR
```

Weirdly when passing the parameter like this, mangohud reports the mode as IMMEDIATE now...

**Comment by flightlessmango** [comment]
I will add some debug info to mangohud so we can see what's going on

**Comment by flightlessmango** [comment]
it's not `--present_mode VK_PRESENT_MODE_MAILBOX_KHR`, you set the int instead
`mangohud vkcube --present_mode 1`
This way mangohud displays the correct present mode

**Comment by shmerl** [comment]
Ah, thanks for the pointer. That indeed works correctly. But then something strange is still happening when forcing the present mode with mesa variable. And it's not working as mailbox in CP2077 by default (even though I think KWin should set it to mailbox).

**Comment by flightlessmango** [comment]
Can you compile mangohud in debug and see what he says about present mode?

**Comment by shmerl** [comment]
Sure, I can try. What flag should I set for Meson to enable it?

**Comment by flightlessmango** [comment]
`--buildtype=debug` should be sufficient

**Comment by shmerl** [comment]
Here is what I see when using debug build:

```
[2024-02-10 20:35:29.876] [MANGOHUD] [debug] [overlay_params.cpp:956] Param: 'present_mode' = '1'
[2024-02-10 20:35:29.876] [MANGOHUD] [debug] [logging.cpp:178] Logger constructed!
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1058] Ram:64940004
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1059] Cpu:AMD Ryzen 9 7950X 16-Core Processor
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1060] Kernel:6.7.4-dirty
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1061] Os:Debian GNU/Linux trixie/sid
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1062] Driver:
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [overlay.cpp:1063] CPU Scheduler:powersave
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [cpu.cpp:506] hwmon: sensor name: iwlwifi_1
[2024-02-10 20:35:29.882] [MANGOHUD] [debug] [cpu.cpp:506] hwmon: sensor name: nct6799
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:538] hwmon: using input: /sys/class/hwmon/hwmon4/temp13_input
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: iwlwifi_1
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: nct6799
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: amdgpu
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: nvme
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: k10temp
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: amdgpu
[2024-02-10 20:35:29.888] [MANGOHUD] [debug] [cpu.cpp:631] hwmon: sensor name: nvme
Selected GPU 0: AMD Radeon RX 7900 XTX (RADV NAVI31), type: DiscreteGpu
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [vulkan.cpp:1853] gpu: AMD Radeon RX 7900 XTX (RADV NAVI31)
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [overlay.cpp:866] drm path check: /sys/class/drm/card1
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [overlay.cpp:883] DeviceID does not match vulkan report 744C
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [overlay.cpp:866] drm path check: /sys/class/drm/card0
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [overlay.cpp:907] Using gpu_metrics of /sys/class/drm/card0/device/gpu_metrics
[2024-02-10 20:35:29.898] [MANGOHUD] [debug] [overlay.cpp:965] Uploading is disabled (permit_upload = 0)
[2024-02-10 20:35:29.901] [MANGOHUD] [debug] [vulkan.cpp:1541] Present mode: FIFO
[2024-02-10 20:35:29.904] [MANGOHUD] [debug] [vulkan.cpp:715] Recreating font image
[2024-02-10 20:35:29.938] [MANGOHUD] [debug] [vulkan.cpp:734] Default font tex size: 1024x2048px
[2024-02-10 20:35:29.945] [MANGOHUD] [debug] [vulkan.cpp:1541] Present mode: FIFO
[2024-02-10 20:35:29.948] [MANGOHUD] [debug] [vulkan.cpp:715] Recreating font image
[2024-02-10 20:35:29.981] [MANGOHUD] [debug] [vulkan.cpp:734] Default font tex size: 1024x2048px
```

Running like this:

```
MESA_VK_WSI_PRESENT_MODE=mailbox MANGOHUD_CONFIG="present_mode" mangohud=true gpu_hud.sh vkcube-wayland
```
(`gpu_hud.sh` is my own script that loads mangohud).

I'm pretty sure that it runs in mailbox, see huge framerate, but mangohud still reports fifo:

![image](https://github.com/flightlessmango/MangoHud/assets/310927/6237d898-48dd-4f15-8fff-1f032eba0681)

**Comment by shmerl** [comment]
I think you should be able to reproduce it pretty easily.

**Comment by flightlessmango** [comment]
It appears that `MESA_VK_WSI_PRESENT_MODE` is setting present_mode after our layer, in that case we're shit out of luck

**Comment by shmerl** [comment]
I see, that makes sense. Then I wonder why CP2077 runs in mailbox mode with Kwin even without forcing it, when vsync is off in the game. I can ping KDE developers about that.

**Comment by flightlessmango** [comment]
> may be you can re-read the mode at some later point 

No it's not possible. The only chance we have is when we intercept the swapchain creation

**Comment by mbriar** [comment]
With VK_EXT_swapchain_maintenance1, the present mode can now be changed on every vkQueuePresentKHR, so would probably make sense to detect that as well.

**Comment by shmerl** [comment]
How does Mesa / radv do it when you force it with `MESA_VK_WSI_PRESENT_MODE`, i.e. will using detection with `VK_EXT_swapchain_maintenance1` detect that too?

**Comment by mbriar** [comment]
> i.e. will using detection with VK_EXT_swapchain_maintenance1 detect that too

no, the mesa env var thing is probably impossible to detect in a layer.

**Comment by shmerl** [comment]
Interestingly, when trying it with Mesa 23.3.5 and Cyberpunk 2077 - it shows immediate. But when trying Mesa-main (git) it shows fifo. I wonder what's going on.

**Comment by flightlessmango** [comment]
looking at `VK_EXT_swapchain_maintenance1` I can't see that we can query present mode, but it sounds like we can set present mode.
I'm leaning towards nuking this options as it doesn't seem like we can know what present mode is used

**Comment by shmerl** [comment]
@mbriar: Any idea how that can be queried?

**Comment by theofficialgman** [comment]
> We're able to check what the app requested for present mode and apply the same logic to that as when we overwrite the present mode. I've written an implementation here that hopefully works as expected [2bc323b](https://github.com/flightlessmango/MangoHud/commit/2bc323b2ed045e38089eca9a9d34d3853e608d3e) and [0ab4c25](https://github.com/flightlessmango/MangoHud/commit/0ab4c25343ba3f61cd738f6d85d577f6f12ce855)
> 
> The opengl part might not be as well as the vulkan one

indeed. the opengl part just shows VSYNC ON at all times for me. the vulkan implementation seems to work fine.

**Comment by shmerl** [comment]
The Vulkan one works fine except for the above cases (forced mode with Mesa and vkd3d-proton scenario).

**Comment by Lassebq** [comment]
I think Vulkan present_mode works great other than the fact it doesn't change with MESA_VK_WSI_PRESENT_MODE. Since you can't query driver directly for current present mode I think the following would be a good solution:
* If the driver is mesa, check the value of MESA_VK_WSI_PRESENT_MODE and if it's set, force HUD to display that mode.
* Otherwise change it based on application preferences

**Comment by Lassebq** [comment]
Not sure what are the recent changes but it seems to display correct present mode when MESA_VK_WSI_PRESENT_MODE is set, except now the framerate isn't limited by refresh rate?

**Comment by shmerl** [comment]
You mean mangohud's framerate limiter isn't working? refresh rate doesn't have to limit framerate otherwise.

**Comment by Lassebq** [comment]
FIFO present mode has to. And no, that issue seems to be irrelevant to mangohud. It's related to this https://gitlab.freedesktop.org/mesa/mesa/-/issues/11379

**Comment by shmerl** [comment]
Yeah, I mean not always has to. But FIFO is old style vsync so yeah. Thanks for the link.

**Comment by LethalManBoob** [comment]
Hello this is still an issue.

**Comment by shmerl** [comment]
I think it can't be solved without some new Vulkan API features for querying the presentation mode. May be propose it for Khronos. It's kind of weird that you can't query it.
